microsoftgraph/msgraph-sdk-python-core
# Feat/telemetry handler

## Overview

Provides a middleware component (`TelemetryHandler`) that attaches metadata to a Graph request in order help the SDK team improve the developer experience.

* Adds a `client-request-id` header with GUID value to request if one is not present.
* Adds `SdkVersion` request header to each request to identify the language and version of the client SDK library(s).
* If available, adds the `featureUsage` value contained in the `RequestContext` to the `SdkVersion` header
* Adds `HostOS` request header to each request to help us identify the OS on which our client SDK is running on.
* Adds a  `RuntimeEnvironment` request header to capture the runtime framework on which the client SDK is running on. 

## Demo

![image](https://user-images.githubusercontent.com/40166690/121528752-6f3ed200-ca04-11eb-9937-5a9ba40ceab2.png)

## Testing Instructions

* Pull the `feat/telemetry-handler` branch
* Instantiate a client and make a call to a graph endpoint
 
```py
client = GraphClient(credential=client_secret_credentials)
resp = client.get(
    '/users',
    params={
        '$select': 'displayName',
        '$top': '10'
    },
)
```
* Observe that the `SdkVersion`, `client-request-id`, `HostOS`, `RuntimeEnvironment` headers were added to the request:
* 
```py
print(resp.request.headers)
```

Fixes #10
